src/storage-types/src/sinks.rs

```diff
@@ -586,6 +687,85 @@ pub enum KafkaSinkAvroFormatState<C: ConnectionAccess = InlinedConnection> {
     },
 }
 
+impl<C: ConnectionAccess> KafkaSinkAvroFormatState<C> {
+    /// Determines if `self` is compatible with another `KafkaSinkAvroFormatState`,
+    /// in such a way that it is possible to turn `self` into `other` through a
+    /// valid series of transformations (e.g. no transformation or `ALTER
+    /// CONNECTION`).
+    pub fn alter_compatible(
+        &self,
+        id: GlobalId,
+        other: &KafkaSinkAvroFormatState<C>,
+    ) -> Result<(), StorageError> {
+        if self == other {
+            return Ok(());
+        }
+
+        match (self, other) {
+            (
+                Self::UnpublishedMaybe {
+                    key_schema,
+                    value_schema,
+                    // Connections may change
+                    csr_connection: _,
+                },
+                Self::UnpublishedMaybe {
+                    key_schema: other_key_schema,
+                    value_schema: other_value_schema,
+                    csr_connection: _,
+                },
+            ) => {
+                let compatibility_checks = [
+                    (key_schema == other_key_schema, "key_schema"),
+                    (value_schema == other_value_schema, "value_schema"),
+                ];
+                for (compatible, field) in compatibility_checks {
+                    if !compatible {
+                        tracing::warn!(
+                    "KafkaSinkAvroFormatState::UnpublishedMaybe incompatible at {field}:\nself:\n{:#?}\n\nother\n{:#?}",
+                    self,
+                    other
+                );
+
+                        return Err(StorageError::IncompatibleConnectionDescriptions { id });
+                    }
```

Comment:
Untested according to coverage info

Reply:
Yep––these being uncovered is expected because these are meant as stopgaps to ensure we don't introduce incompatible changes